lib/finitomata/transition.ex

```diff
@@ -293,26 +303,148 @@ defmodule Finitomata.Transition do
   @spec loops(:states | :transitions, [t()]) :: Enumerable.t(t()) | [Path.t()]
   def loops(what \\ :states, transitions)
 
+  def loops(:states, transitions) do
+    :transitions
+    |> loops(transitions)
+    |> to_path()
+  end
+
   def loops(:transitions, transitions) do
     transitions
     |> states(true)
     |> Stream.flat_map(&do_loop(&1, transitions, [], []))
   end
 
-  def loops(:states, transitions) do
-    :transitions
-    |> loops(transitions)
-    |> to_path()
+  defp do_loop(state, transitions, path, paths) do
+    count = 2 + div(24, count(:transitions, transitions))
```

Comment:
Nah. In the worst cases, we are after calculating _all_ the permutations between _all_ the states. Which leads to infinite times. This math is the result of me experimenting with numbers. I have a degree in physics after all, so that’s a made-up constant, let’s call it Aleksei’s parameter.